### ISSUE TITLE
handsomegiraffe - [M-01] Incorrect refund of execution fee to user

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-02-gmx-judging#212]
handsomegiraffe

medium

# [M-01] Incorrect refund of execution fee to user

## Summary
During execution of Deposits, Withdrawals and Orders, users are refunded part of the `executionFee` after accounting for `gasUsed` during the transaction. In the codebase, an incorrect value of `startingGas` is used to calculate the `gasUsed`, resulting in users getting less than what they should be refunded. 

## Vulnerability Detail
Vulnerability exists in DepositHandler.sol, WithdrawalHandler.sol and OrderHandler.sol. Using DepositHandler.sol as an example:

![image](https://user-images.githubusercontent.com/83704326/227539224-f59d4cb7-f638-4ddd-a7b3-7949e8fef1b8.png)

((1) In [line 94](https://github.com/gmx-io/gmx-synthetics/blob/b1557fa286c35f54c65a38a7b57baf87ecad1b5b/contracts/exchange/DepositHandler.sol#L94) of DepositHandler.sol, Order Keepers call `executeDeposit()` and `startingGas` is forwarded to an external call `_executeDeposit`. 
(2) In ExecuteDepositUtils.sol, `_executeDeposit` further calls `GasUtils.payExecutionFee(... params.startingGas`. 
(3) Then in GasUtils.sol, `payExecutionFee()` calculates `gasUsed = startingGas - gasleft();` 
(4) `gasUsed` is used to calculate `executionFeeForKeeper`, and after paying the fee to keeper, the remainder of `executionFee` (previously paid by user) is refunded to the user

The issue lies with (1) where `startingGas` is passed into `_executeDeposit` and assumed to be all remaining gas left. [EIP-150](https://eips.ethereum.org/EIPS/eip-150) defines the "all but one 64th" rule, which states that always at least 1/64 of the gas still not used for this transaction cannot be sent along. Therefore, in (3) `gasUsed` is overstated by 1/64 and the refund back to user in (4) is incorrect (less than what user should get back). 

**Proof of Concept**
![image](https://user-images.githubusercontent.com/83704326/227538387-f2f2ca87-d784-49a5-b043-d779a92fcff8.png)

In the test above, it is demonstrated that external function calls are forwarded with only 63/64 of the remaining gas. A separate internal function call used to demonstrate the difference in gas costs.  

## Impact
GMX Users will receive an incorrect refund from the execution fee and will be overpaying for deposit, withdraw and order executions.

## Code Snippet
https://github.com/gmx-io/gmx-synthetics/blob/b1557fa286c35f54c65a38a7b57baf87ecad1b5b/contracts/exchange/DepositHandler.sol#L100
https://github.com/gmx-io/gmx-synthetics/blob/b1557fa286c35f54c65a38a7b57baf87ecad1b5b/contracts/exchange/WithdrawalHandler.sol#L130
https://github.com/gmx-io/gmx-synthetics/blob/b1557fa286c35f54c65a38a7b57baf87ecad1b5b/contracts/exchange/OrderHandler.sol#L174

## Tool used
Hardhat
Manual Review

## Recommendation

In DepositHandler.sol, for `executeDeposit` it is recommended that `startingGas()` is calculated **after** the external call is made.
![image](https://user-images.githubusercontent.com/83704326/227540020-c570a91b-4bdd-41be-9525-e069bab895f6.png)

Alternatively, in GasUtils.sol, gasUsed could be computed with 63/64 of startingGas, in order to obtain the correct refund amount to the user. This would also apply to Withdraw and Order executions which have similar code flows.
![image](https://user-images.githubusercontent.com/83704326/227539740-4df5497a-709d-4e4c-ad00-ab492fc5b74c.png)

**Comment by xvi10** [comment]
this is a valid concern but we do not think this requires a change in the contracts, startingGas is measured after withOraclePrices, so there will be some extra gas consumed for that call, additionally the keeper should be incentivised with a small fee to execute the requests, the value for Keys.EXECUTION_GAS_FEE_BASE_AMOUNT can be adjusted to account for these

**Comment by IllIllI000** [comment]
@hrishibhat final decision up to you: https://discord.com/channels/812037309376495636/1089859477412520006/1090901782147891321

**Comment by Jiaren-tang** [comment]
Escalate for 10 USDC. I do not recommend escalating this issue for a reward of 10 USDC, as the impact is too trivial. According to the protocol's comments, not forwarding 1/64th of the gas to compensate for the execution fee does not block order flow execution, and Keys.EXECUTION_GAS_FEE_BASE_AMOUNT can be configured to easily solve this issue.

Furthermore, based on Sherlock's severity guide:

https://docs.sherlock.xyz/audits/judging/judging

> a Medium severity rating should only be given if there is a viable scenario (even if unlikely) that could cause the protocol to enter a state where a material amount of funds can be lost. The attack path is possible with assumptions that either mimic on-chain conditions or reflect conditions that have a reasonable chance of becoming true in the future. The more expensive the attack is for an attacker, the less likely it will be included as a Medium (holding all other factors constant). The vulnerability must be something that is not considered an acceptable risk by a reasonable protocol team.

Based on the protocol's comment, this issue is clearly considered an acceptable risk by a reasonable protocol team. Therefore, I do not believe it meets the criteria for a Medium severity rating.
@xvi10 @IllIllI000

**Comment by IllIllI000** [comment]
I agree with the [second](https://github.com/sherlock-audit/2023-02-gmx-judging/issues/212#issuecomment-1495786878) escalation for its stated reasons. In addition, if the swap path is long enough, the base fee may not be large enough to cover the number of swaps

**Comment by hrishibhat** [comment]
Escalation accepted

Considering this issue as valid based on Sponsor, 2nd Escalation and Lead Watson comments.

**Comment by xvi10** [comment]
Fix in https://github.com/gmx-io/gmx-synthetics/pull/119
